compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -784,25 +786,31 @@ class ClassfileParser(
       if (scan(tpnme.TASTYATTR)) {
         val attrLen = in.nextInt
         if (attrLen == 0) { // A tasty attribute implies the existence of the .tasty file
-          def readTastyForClass(jpath: nio.file.Path): Array[Byte] = {
-            val plainFile = new PlainFile(io.File(jpath).changeExtension("tasty"))
-            if (plainFile.exists) plainFile.toByteArray
-            else {
-              ctx.error("Could not find " + plainFile)
-              Array.empty
-            }
-          }
-          val tastyBytes = classfile.underlyingSource match { // TODO: simplify when #3552 is fixed
+          val tastyBytes: Array[Byte] = classfile.underlyingSource match { // TODO: simplify when #3552 is fixed
             case None =>
               ctx.error("Could not load TASTY from .tasty for virtual file " + classfile)
-              Array.empty[Byte]
+              Array.empty
             case Some(jar: ZipArchive) => // We are in a jar
-              val jarFile = JarArchive.open(io.File(jar.jpath))
-              readTastyForClass(jarFile.jpath.resolve(classfile.path))
-              // Do not close the file system as some else might use it later. Once closed it cannot be re-opened.
-              // TODO find a way to safly close the file system or ose some other abstraction
+              val cl = new URLClassLoader(Array(jar.jpath.toUri.toURL))
+              val path = classfile.path.stripSuffix(".class") + ".tasty"
+              val stream = cl.getResourceAsStream(path)
+              if (stream != null) {
+                val tasty = Array.newBuilder[Byte]
+                tasty.sizeHint(stream.available())
+                while (stream.available() > 0) // TODO improve performance
```

Comment:
I think the proper way to read from an `InputStream`, is to `read()` until `-1` is returned

Reply:
Changed it